utils/cloud_metadata.py

```diff
@@ -218,6 +238,31 @@ def get_metadata(agentConfig):
                 log.debug("Collecting EC2 Metadata failed %s", str(e))
                 pass
 
+        # Only check if we don't know yet or if it is known to be OpenStack
+        if EC2.is_openstack is None or EC2.is_openstack is True:
+            log.info(u"Attempting to get OpenStack meta_data.json")
+            openstack_metadata_url = EC2.EC2_METADATA_HOST + "/openstack/latest/meta_data.json"
+            try:
+                r = requests.get(openstack_metadata_url, timeout=EC2.TIMEOUT)
+                r.raise_for_status() # Fail on 404 etc
+
+                EC2.is_openstack = True
+                openstack_metadata = r.json()
+                # Set the "also known as" metadata similar to AWS EC2
+                EC2.metadata['host_aliases'] = [
+                    openstack_metadata['uuid'],
+                    openstack_metadata['name'],
+                ]
+
+                # the OpenStack instance-id can be recycled but the uuid will always be unique
+                if agentConfig['openstack_use_uuid']:
+                    log.info(u"Using the instance's uuid for instance-id")
+                    EC2.metadata["instance-id"] = openstack_metadata['uuid']
```

Comment:
It might make sense to send this up as another host alias, instead of overwriting the ec2 instance id. Instance id's can be important to people for various reasons, as it's a stable (external) identifier, Amazon knows what to do with it in the case of a support ticket, and so on. I don't think it's a good idea to overwrite it.

Instead of you could just make a new metadata piece that sends it up as a separate host alias, setting it as the main hostname if this is enabled.